### PR TITLE
Compiles against cuDNNv6 with new support for dilated convolution.

### PR DESCRIPTION
**Pull request by DickJC123**

### CODE REVIEW

File: src/operator/cudnn_deconvolution-inl.h [hunk 1]
```diff
@@ -263,6 +264,17 @@ class CuDNNDeconvolutionOp : public Operator {
                                                1,
                                                1,
                                                CUDNN_CROSS_CORRELATION), CUDNN_STATUS_SUCCESS);
+      #else
+      CHECK_EQ(cudnnSetConvolution2dDescriptor(conv_desc_,
+                                               pad_y,
+                                               pad_x,
+                                               param_.stride[0],
+                                               param_.stride[1],
+                                               1,
```
Comment (piiswrong) [inline src/operator/cudnn_deconvolution-inl.h]: we can also support dilation for deconv right?

Comment (DickJC123) [inline src/operator/cudnn_deconvolution-inl.h]: Sorry for the delayed response.  Just starting to look at the existing deconvolution code.  Looks like dilation support is not present along the non-cuDNN path as well. This PR doesn't touch deconvolution.

**Comment by piiswrong** [comment]
Looks like it's segfaulting: http://ec2-52-25-96-65.us-west-2.compute.amazonaws.com/blue/organizations/jenkins/mxnet/detail/PR-5606/2/pipeline/234

File: tests/python/gpu/test_operator_gpu.py [hunk 2]
```diff
@@ -358,19 +375,78 @@ def test_pooling_with_type():
...
+    # Stride > 1
+    sym = mx.sym.Deconvolution(num_filter=2, kernel=(3,3), stride=(2,2), name='deconv')
+    sym_no_cudnn = mx.sym.Deconvolution(num_filter=2, kernel=(3,3), stride=(2,2), cudnn_off=True, name='deconv')
+    check_consistency_NxM([sym, sym_no_cudnn], ctx_list)
+    # Dilate > 1
+    sym = mx.sym.Deconvolution(num_filter=2, kernel=(3,3), dilate=(2,2), name='deconv')
+    sym_no_cudnn = mx.sym.Deconvolution(num_filter=2, kernel=(3,3), dilate=(2,2), cudnn_off=True, name='deconv')
+    check_consistency_NxM([sym, sym_no_cudnn], ctx_list)
+
+#    # 3D convolution (not yet enabled)
```
Comment (piiswrong) [inline tests/python/gpu/test_operator_gpu.py]: why not enabled?

Comment (DickJC123) [inline tests/python/gpu/test_operator_gpu.py]: Thanks for pointing out the typo: it should say that 1D/3D **de**convolutions are not yet enabled.  Did you want to delay the cuDNNv6 PR to wait for uniform support of 1D/3D deconvolutions?  Currently at deconvolution-inl.h line 138:

LOG(FATAL) << "If not using CUDNN only 2D-Deconvolution is supported.";

File: src/operator/convolution.cu [hunk 3]
```diff
@@ -28,15 +28,63 @@ Operator* CreateOp<gpu>(ConvolutionParam param, int dtype,
...
-  } else {
-    MSHADOW_REAL_TYPE_SWITCH(dtype, DType, {
+  // The NVIDIA Pascal architecture was the first to include 16-bit ALUs.
+  // Thus, when the framework is compiled with MSHADOW_USE_PASCAL == 1, we
+  // perform the convolution calculation in 16-bit when the tensor type is
+  // also 16-bit.  For NVIDIA architectures earlier than Pascal (so Maxwell
+  // and Kepler), the computation precision is always at least 32-bits.
+#if MSHADOW_USE_PASCAL == 1
+  // true fp16
+  int desired_forward_compute_type = dtype;
```
Comment (piiswrong) [inline src/operator/convolution.cu]: @DickJC123 can we detect this in runtime so we don't need to provide separate binaries for pascal/maxwell gpu?

**Pull request merged**